### PR TITLE
Fix gradient accumulation with small dataset sizes

### PR DESCRIPTION
**Pull request by nelson-liu**
Right now, `step` is reset across epochs. This is problematic if, say, you're training on a dataset with 8 examples but have a per-gpu batch size of 8 and want to do gradient accumulation for 4 steps. You'll never reach `step = 4`, and then never do gradient updates.

this fix just persists `step` across epochs

### CODE REVIEW

**Comment by timoschick** [comment]
Well spotted, thanks!
